// openveo-api/lib/applicationStorage.js
"use strict"

// Stores a bunch of information for all the application
var plugins, menu, database;

/**
 * Gets the list of loaded openveo plugins.
 * @return Array The list of loaded plugins
 */
module.exports.getPlugins = function(){
  return plugins;
};

/**
 * Sets the list of openveo plugins.
 * @param Array subPlugins The list of plugins
 */
module.exports.setPlugins = function(subPlugins){
  plugins = subPlugins;
};

/**
 * Gets the computed back office menu with all links.
 * @return Array The list of back office links
 */
module.exports.getMenu = function(){
  return menu;
};

/**
 * Sets the back office menu list of links.
 * @param Array newMenu The list of back office menu links
 */
module.exports.setMenu = function(newMenu){
  menu = newMenu;
};

/**
 * Gets the database.
 * @return Database A Database object
 */
module.exports.getDatabase = function(){
  return database;
};

/**
 * Sets the database.
 * @param Database newDatabase The new database of the application
 */
module.exports.setDatabase = function(newDatabase){
  database = newDatabase;
};
