// @openveo/official-plugin-example/conf.js
'use strict';

module.exports = {
  http: {
    routes: {
      public: {
        'get /test': 'ExampleController.getTestAction',
        'post /test': 'ExampleController.postTestAction',
        'put /test': 'ExampleController.putTestAction',
        'DELETE /test': 'ExampleController.deleteTestAction',
        '/': 'ExampleController.homeAction',
        '/test': 'invalid action'
      },
      private: {
        'get /test': 'AdminExampleController.getTestAction',
        'post /test': 'AdminExampleController.postTestAction',
        'put /test': 'AdminExampleController.putTestAction',
        'DELETE /test': 'AdminExampleController.deleteTestAction',
        '/': 'ExampleController.invalidAction',
        '/test': 'invalid action',
        '*': [
          'AdminExampleController.allFirstAction',
          'AdminExampleController.allSecondAction'
        ]
      },
      ws: {
        'get /test': 'WSExampleController.getTestAction',
        'post /test': 'WSExampleController.postTestAction',
        'put /test': 'WSExampleController.putTestAction',
        'DELETE /test': 'WSExampleController.deleteTestAction',
        '*': [
          'WSExampleController.allFirstAction',
          'WSExampleController.allSecondAction'
        ]
      }
    }
  },
  socket: {
    namespaces: {
      test: {
        public: {
          test1: 'NamespaceExampleController.test1Action'
        },
        private: {
          test2: 'NamespaceExampleController.test2Action'
        }
      }
    }
  },
  entities: {
    'entity-example': 'EntityExampleController'
  },
  webServiceScopes: {
    scope: {
      name: 'Scope name',
      description: 'Scope description',
      paths: [
        'get /example/test',
        'post /example/test',
        'put /example/test'
      ]
    }
  },
  permissions: [
    {
      id: 'permission',
      name: 'Permission name',
      description: 'Permission description',
      paths: [
        'get /example/test',
        'post /example/test',
        'put /example/test'
      ]
    }
  ],
  backOffice: {
    menu: [],
    scriptLibFiles: {},
    scriptFiles: {
      dev: [],
      prod: []
    },
    cssFiles: []
  },
  libraries: []
};
