// openveo-api/lib/database/mongodb/MongoDatabase.js
"use scrict"

// Module dependencies
var util = require("util");
var mongodb = require("mongodb");
var Database = require("../../Database.js");
var MongoClient = mongodb.MongoClient;

/**
 * Creates a MongoDatabase instance.
 * @param Object databaseConf A database configuration object like
 * {
 *   "type" : "mongodb",
 *   "host" : "localhost",
 *   "port" : 27017,
 *   "database" : "openveo",
 *   "username" : "openveo",
 *   "password" : "openveo"
 * }
 */
function MongoDatabase(databaseConf){
  Database.call(this, databaseConf);
}

module.exports = MongoDatabase;
util.inherits(MongoDatabase, Database);

/**
 * Establishes connection to the database.
 * @param Function callback The function to call when connection to the
 * database is done
 *   - Error The error if an error occurred, null otherwise 
 */
MongoDatabase.prototype.connect = function(callback){
  var self = this;
  
  var connectionUrl = "mongodb://" + this.conf.username + ":" + this.conf.username + "@" + this.conf.host + ":" + this.conf.port + "/" + this.conf.database;
  var mongoClient = MongoClient.connect(connectionUrl, function(error, db){
    
    // Connection failed
    if(error)
      callback(error);
    
    // Connection succeeded
    else{
      self.db = db;
      callback();
    }
  });
 
};

/**
 * Inserts a document into a collection.
 * @param String collection The collection to work on
 * @param Objet data The document to insert into the collection
 * e.g.
 * {
 *   id : 25,
 *   status : 1,
 *   type : "type"
 * }
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise 
 */
MongoDatabase.prototype.insert = function(collection, data, callback){
  var collection = this.db.collection(collection);
  collection.insert(data, callback);
};

/**
 * Removes a document from a collection.
 * @param String collection The collection to work on
 * @param Object criteria The remove criteria
 * e.g.
 * {
 *   id : 5
 * }
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
MongoDatabase.prototype.remove = function(collection, criteria, callback){
  if(criteria && Object.keys(criteria).length){
    var collection = this.db.collection(collection);
    collection.remove(criteria, callback);
  }
};

/**
 * Updates a document.
 * @param String collection The collection to work on
 * @param Object criteria The update criteria
 * e.g.
 * {
 *   id : 5
 * }
 * @param Object data Data to update
 * e.g.
 * {
 *   type : "new type"
 * }
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise 
 */
MongoDatabase.prototype.update = function(collection, criteria, data, callback){
  var collection = this.db.collection(collection);
  collection.update(criteria, {$set : data}, callback);
};

/**
 * Gets a list of documents.
 * @param String collection The collection to work on
 * @param Object criteria An object of criterias
 * e.g.
 * {
 *   type: "vimeo",
 *   id: { $gte: 100 }
 * }
 * @param Object projection Fields to return using projection operators
 * e.g.
 * {
 *   _id : 0,
 *   status : 1
 * }
 * @param Number limit An optional limit number of items to retrieve
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise 
 *   - Array The retrieved data
 */
MongoDatabase.prototype.get = function(collection, criteria, projection, limit, callback){
  var collection = this.db.collection(collection);

  var criteria = criteria || {};
  var projection = projection || {};
  var limit = limit || -1;

  if(limit === -1)
    collection.find(criteria, projection).toArray(callback);
  else
    collection.find(criteria, projection).limit(limit).toArray(callback);
};
