// @openveo/example/conf.js
'use strict';

module.exports = {
  routes: {
    public: {
      'get /test': 'exampleController.getTestAction',
      'post /test': 'exampleController.postTestAction',
      'put /test': 'exampleController.putTestAction',
      'DELETE /test': 'exampleController.deleteTestAction',
      '/': 'exampleController.homeAction',
      '/test': 'invalid action'
    },
    private: {
      'get /test': 'adminExampleController.getTestAction',
      'post /test': 'adminExampleController.postTestAction',
      'put /test': 'adminExampleController.putTestAction',
      'DELETE /test': 'adminExampleController.deleteTestAction',
      '/': 'exampleController.invalidAction',
      '/test': 'invalid action',
      '*': [
        'adminExampleController.allFirstAction',
        'adminExampleController.allSecondAction'
      ]
    },
    ws: {
      'get /test': 'wsExampleController.getTestAction',
      'post /test': 'wsExampleController.postTestAction',
      'put /test': 'wsExampleController.putTestAction',
      'DELETE /test': 'wsExampleController.deleteTestAction',
      '*': [
        'wsExampleController.allFirstAction',
        'wsExampleController.allSecondAction'
      ]
    }
  },
  entities: {
    'entity-example': 'EntityExample'
  },
  webServiceScopes: {
    scope: {
      name: 'Scope name',
      description: 'Scope description',
      paths: [
        'get /test',
        'post /test',
        'put /test'
      ]
    }
  },
  permissions: [
    {
      id: 'permission',
      name: 'Permission name',
      description: 'Permission description',
      paths: [
        'get /test',
        'post /test',
        'put /test'
      ]
    }
  ],
  backOffice: {
    menu: [],
    scriptLibFiles: {},
    scriptFiles: {
      dev: [],
      prod: []
    },
    cssFiles: []
  }
};
