// openveo-api/lib/providers/VideoProvider.js
"use strict"

// The name of the video collection
var VIDEO_COLLECTION = "videos";

/**
 * Creates a VideoProvider to interact with database videos collection.
 * @param Database database The database to interact with
 */
function VideoProvider(database){
  this.database = database;
  this.pendingUpdateOperations;
  
  if(!this.database)
    throw new Error("VideoProvider needs a database");
}

module.exports = VideoProvider;

// Error codes
VideoProvider.COPY_ERROR = 0,
VideoProvider.UNLINK_ERROR = 1;
VideoProvider.PACKAGE_NOT_DEFINED_ERROR = 2;
VideoProvider.EXTRACT_ERROR = 3;
VideoProvider.VALIDATION_ERROR = 4;
VideoProvider.CREATE_VIDEO_PUBLIC_DIR_ERROR = 5;
VideoProvider.SAVE_VIDEO_DATA_ERROR = 6;
VideoProvider.SAVE_TIMECODE_ERROR = 7;
VideoProvider.UPLOAD_ERROR = 8;
VideoProvider.SCAN_FOR_IMAGES_ERROR = 9;
VideoProvider.CREATE_VIDEOS_PUBLIC_DIR_ERROR = 10;

// Status codes
VideoProvider.ERROR_STATUS = 0;
VideoProvider.SUCCESS_STATUS = 1;
VideoProvider.PENDING_STATUS = 2;

// States codes
VideoProvider.PENDING_STATE = 0;
VideoProvider.COPYING_STATE = 1;
VideoProvider.EXTRACTING_STATE = 2;
VideoProvider.VALIDATING_STATE = 3;
VideoProvider.PREPARING_STATE = 4;
VideoProvider.SENDING_STATE = 5;
VideoProvider.SENT_STATE = 6;
VideoProvider.PUBLISHED_STATE = 7;
VideoProvider.ERROR_STATE = 8;

/**
 * Adds a new video into database.
 * @param Object videoPackage Information about the video
 *   {
 *     "id" : 1422731934859,
 *     "status" : 1,
 *     "properties" : [],
 *     "published" : false,
 *     "type" : "vimeo",
 *     "path" : "C:/Temp/",
 *     "originalPackagePath" : "C:/Temp/video-package.tar",
 *     "packagePath" : "E:/openveo/node_modules/openveo-publish/tmp/1422731934859.tar",
 *     "metadata" : {
 *       "profile": "2",
 *       "audio-input": "analog-top",
 *       "date": "13/01/1970 20:36:15",
 *       "format": "mix-pip",
 *       "rich-media": true,
 *       "profile-settings": {
 *         "video-bitrate": 1000000,
 *         "id": "2",
 *         "video-height": 720,
 *         "audio-bitrate": 128000,
 *         "name": "Haute définition"
 *       },
 *       "id": "1970-01-13_20-36-15",
 *       "format-settings": {
 *         "source": "mix-raw",
 *         "id": "mix-pip",
 *         "name": "Mélangé caméra incrustée",
 *         "template": "pip"
 *       },
 *       "date-epoch": 1107375,
 *       "storage-directory": "/data/1970-01-13_20-36-15",
 *       "filename": "video.mp4",
 *       "duration": 20
 *     }
 *   }
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise 
 */
VideoProvider.prototype.add = function(videoPackage, callback){
  this.database.insert(VIDEO_COLLECTION, {
    id : videoPackage.id, 
    status : videoPackage.status,
    metadata : videoPackage.metadata,
    url : videoPackage.url,
    type : videoPackage.type,
    errorCode : videoPackage.errorCode,
    published : videoPackage.published,
    properties : videoPackage.properties
  }, callback);
};

/**
 * Updates video state.
 * @param Number id The id of the video to update
 * @param String state The state of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateState = function(id, state, callback){
  updateVideoProperty.call(this, id, "state", state, callback);
};

/**
 * Updates video status.
 * @param Number id The id of the video to update
 * @param String status The status of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateStatus = function(id, status, callback){
  updateVideoProperty.call(this, id, "status", status, callback);
};

/**
 * Updates video error code.
 * @param Number id The id of the video to update
 * @param Number errorCode The error code of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateErrorCode = function(id, errorCode, callback){
  updateVideoProperty.call(this, id, "errorCode", errorCode, callback);
};

/**
 * Updates video link.
 * @param Number id The id of the video to update
 * @param String link The link of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateLink = function(id, link, callback){
  updateVideoProperty.call(this, id, "link", link, callback);
};

/**
 * Updates video id for video platform.
 * @param String id The id of the video to update
 * @param String idVideoPlatform The id of the video in the video platform
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateVideoId = function(id, idVideoPlatform, callback){
  updateVideoProperty.call(this, id, "videoId", idVideoPlatform, callback);
};

/**
 * Updates video metadata for video platform.
 * @param Number id The id of the video to update
 * @param String metadata The metadata of the video in the video platform
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateMetadata = function(id, metadata, callback){
  updateVideoProperty.call(this, id, "metadata", metadata, callback);
};

/**
 * Gets the list of videos.
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 *   - Array The list of videos as objects
 */
VideoProvider.prototype.getVideos = function(callback){
  this.database.get(VIDEO_COLLECTION, null, null, -1, callback);
};

/**
 * Gets a video by its id.
 * @param String id The id of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 *   - Object The video
 */
VideoProvider.prototype.getVideo = function(id, callback){
  this.database.get(VIDEO_COLLECTION, {id : parseInt(id)}, null, 1, function(error, videos){
    if(videos && videos.length)
      callback(error, videos[0]);
    else
      callback(error);
  });
};

/**
 * Removes a video by its id.
 * @param String id The id of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.removeVideo = function(id, callback){
  this.database.remove(VIDEO_COLLECTION, {id : parseInt(id)}, callback);
};

/**
 * Updates video information.
 * @param String id The id of the video
 * @param Object info The video info
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.updateVideo = function(id, info, callback){
  this.database.update(VIDEO_COLLECTION, {id : parseInt(id)}, info, callback);
};

/**
 * Publishes a video.
 * Change the state of the video to published only if its state if 
 * actually sent.
 * @param String id The id of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.publishVideo = function(id, callback){
  this.database.update(VIDEO_COLLECTION, {id : parseInt(id), state : VideoProvider.SENT_STATE}, {state : VideoProvider.PUBLISHED_STATE}, callback);
};

/**
 * Unpublishes a video.
 * Change the state of the video to sent only if its state if
 * actually published.
 * @param String id The id of the video
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise
 */
VideoProvider.prototype.unpublishVideo = function(id, callback){
  this.database.update(VIDEO_COLLECTION, {id : parseInt(id), state : VideoProvider.PUBLISHED_STATE}, {state : VideoProvider.SENT_STATE}, callback);
};

/**
 * Updates the property of a given video.
 * Update operations are grouped by event loop and only one request 
 * is made to the database for each video id.
 * @param Number videoId The id of the video to update
 * @param String propertyName The name of the property to update
 * @param PrimitiveValue propertyValue The value of the property 
 */
function updateVideoProperty(videoId, propertyName, propertyValue, callback){
  var self = this;
  
  // No pending update operations for now
  if(!this.pendingUpdateOperations){
    this.pendingUpdateOperations = {};
    
    process.nextTick(function(){
      for(var videoId in self.pendingUpdateOperations){
        self.database.update(VIDEO_COLLECTION, {
            id : parseInt(videoId)
          }, self.pendingUpdateOperations[videoId], function(error){
            if(callback)
              callback(error);
        });
      }
      self.pendingUpdateOperations = null;
    });
  }
  
  // Add update opration to pending operations
  if(!this.pendingUpdateOperations[videoId])
    this.pendingUpdateOperations[videoId] = {};

  this.pendingUpdateOperations[videoId][propertyName] = propertyValue;
}

/**
 * Updates a list of properties for the given video.
 *
 * e.g.
 * updateVideoProperties("13545", {
 *   "link" : "/publish/video/13545",
 *   "errorCode" : 2
 * });
 *
 * @param Number id The id of the video to update
 * @param Object properties A key value of properties to update
 * @param Function callback The function to call when it's done
 *   - Error The error if an error occurred, null otherwise 
 */
function updateVideoProperties(id, properties, callback){
  if(properties && typeof properties === "object"){
    this.database.update(VIDEO_COLLECTION, {
      id : id
    },properties, function(error){
      if(callback)
        callback(error);
    });
  }
}
